pyro/ops/welford.py

```diff
@@ -0,0 +1,70 @@
+from collections import OrderedDict
+
+import torch
+
+
+class WelfordCovariance(object):
+    """
+    Implements Welford's online scheme for estimating (co)variance (see :math:`[1]`).
+    Useful for adapting diagonal and dense mass structures for HMC.
+
+    **References**
+
+    [1] `The Art of Computer Programming`,
+    Donald E. Knuth
+    """
+    def __init__(self, diagonal=True):
+        self.diagonal = diagonal
+        self.reset()
+
+    def _deltas(self, z):
+        deltas = OrderedDict()
+        for site_name, site_value in z.items():
+            mean = self.means.get(site_name, 0.)
+            delta_pre = site_value - mean
+            updated_mean = mean + delta_pre / self.n_samples
+            delta_post = site_value - updated_mean
+            deltas[site_name] = (updated_mean, delta_pre, delta_post)
+        return deltas
+
+    def _unroll(self, z):
+        unrolled = OrderedDict()
+        for site_name, rolled in z.items():
+            for i, x in enumerate(rolled.view(-1)):
+                unrolled["{}_{}".format(site_name, i)] = x
+        return unrolled
+
+    def reset(self):
+        self.means = OrderedDict()
+        self.variances = OrderedDict()
+        self.n_samples = 0
+
+    def update(self, z):
+        self.n_samples += 1
+        deltas = self._deltas(self._unroll(z))
+        for i, (site_x, (mean_x, delta_x_pre, delta_x_post)) in enumerate(deltas.items()):
+            self.means[site_x] = mean_x
+            for j, (site_y, (mean_y, delta_y_pre, delta_y_post)) in enumerate(deltas.items()):
+                # Only compute the upper triangular covariance.
+                if i == j or (i < j and not self.diagonal):
+                    self.variances[(site_x, site_y)] = self.variances.get((site_x, site_y), 0.) + \
+                                                       delta_x_pre * delta_y_post
```

Comment:
I think this will work for scalars but not for tensors of different shape. I think you'll need to keep track of outer products via something like
```py
self.covariances[site_x, site_y] += delta_x_pre.unsqueeze(-1) * delta_y_post
```
Alternatively, you could `torch.cat` everything into a single tensor, as we do in pyro.contrib.autoguide.

Reply:
```python
self.variances[site_x, site_y] += delta_x_pre * delta_y_post
```
I think the above is actually ok in this case because we're iterating over the deltas (which have already been unrolled) so both x and y are always scalars in this case. Unless I've misunderstood?